In [159]:
import numpy as np 
import pandas as pd 

In [160]:
df = pd.read_csv('D:/Study/MEC/Tasks/DS/Datasets/UCL_Finals_1955-2023.csv')
df

Season  Country          Winners Score           Runners-up  \
0   1955–56    Spain      Real Madrid   4–3                Reims   
1   1956–57    Spain      Real Madrid   2–0           Fiorentina   
2   1957–58    Spain      Real Madrid   3–2                Milan   
3   1958–59    Spain      Real Madrid   2–0                Reims   
4   1959–60    Spain      Real Madrid   7–3  Eintracht Frankfurt   
..      ...      ...              ...   ...                  ...   
64  2018–19  England        Liverpool   2–0    Tottenham Hotspur   
65  2019–20  Germany    Bayern Munich   1–0  Paris Saint-Germain   
66  2020–21  England          Chelsea   1–0      Manchester City   
67  2021–22    Spain      Real Madrid   1–0            Liverpool   
68  2022–23  England  Manchester City   1–0          Inter Milan   

       Country.1                                      Venue Attend­ance  \
0         France            Parc des Princes, Paris, France      38,239   
1          Italy           Santiago Bernabéu, Madrid, Spain     124,000   
2          Italy          Heysel Stadium, Brussels, Belgium      67,000   
3         France     Neckarstadion, Stuttgart, West Germany      72,000   
4   West Germany            Hampden Park, Glasgow, Scotland     127,621   
..           ...                                        ...         ...   
64       England       Metropolitano Stadium, Madrid, Spain      63,272   
65        France           Estádio da Luz, Lisbon, Portugal           0   
66       England         Estádio do Dragão, Porto, Portugal      14,110   
67       England       Stade de France, Saint-Denis, France      75,000   
68         Italy  Atatürk Olympic Stadium, Istanbul, Turkey      71,412   

                              Notes  
0                               NaN  
1                               NaN  
2   Match was won during extra time  
3                               NaN  
4                               NaN  
..                              ...  
64                              NaN  
65                              NaN  
66                              NaN  
67                              NaN  
68                              NaN  

[69 rows x 9 columns]

In [161]:
df2 = {'Season': '2023-24', 'Country': 'Spain', 'Winners': 'Real Madrid','Score':'2–0','Runners-up':'Borussia Dortmund','Country.1':'Germany','Venue':'Wembley Stadium, London, England','Attend­ance':'86,212',}

In [162]:
df = df._append(df2, ignore_index = True)

In [163]:
df

Season  Country          Winners Score           Runners-up  \
0   1955–56    Spain      Real Madrid   4–3                Reims   
1   1956–57    Spain      Real Madrid   2–0           Fiorentina   
2   1957–58    Spain      Real Madrid   3–2                Milan   
3   1958–59    Spain      Real Madrid   2–0                Reims   
4   1959–60    Spain      Real Madrid   7–3  Eintracht Frankfurt   
..      ...      ...              ...   ...                  ...   
65  2019–20  Germany    Bayern Munich   1–0  Paris Saint-Germain   
66  2020–21  England          Chelsea   1–0      Manchester City   
67  2021–22    Spain      Real Madrid   1–0            Liverpool   
68  2022–23  England  Manchester City   1–0          Inter Milan   
69  2023-24    Spain      Real Madrid   2–0    Borussia Dortmund   

       Country.1                                      Venue Attend­ance  \
0         France            Parc des Princes, Paris, France      38,239   
1          Italy           Santiago Bernabéu, Madrid, Spain     124,000   
2          Italy          Heysel Stadium, Brussels, Belgium      67,000   
3         France     Neckarstadion, Stuttgart, West Germany      72,000   
4   West Germany            Hampden Park, Glasgow, Scotland     127,621   
..           ...                                        ...         ...   
65        France           Estádio da Luz, Lisbon, Portugal           0   
66       England         Estádio do Dragão, Porto, Portugal      14,110   
67       England       Stade de France, Saint-Denis, France      75,000   
68         Italy  Atatürk Olympic Stadium, Istanbul, Turkey      71,412   
69       Germany           Wembley Stadium, London, England      86,212   

                              Notes  
0                               NaN  
1                               NaN  
2   Match was won during extra time  
3                               NaN  
4                               NaN  
..                              ...  
65                              NaN  
66                              NaN  
67                              NaN  
68                              NaN  
69                              NaN  

[70 rows x 9 columns]

In [164]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Season       70 non-null     object
 1   Country      70 non-null     object
 2   Winners      70 non-null     object
 3   Score        70 non-null     object
 4   Runners-up   70 non-null     object
 5   Country.1    70 non-null     object
 6   Venue        70 non-null     object
 7   Attend­ance  70 non-null     object
 8   Notes        18 non-null     object
dtypes: object(9)
memory usage: 5.1+ KB


In [165]:
df.columns

Index(['Season', 'Country', 'Winners', 'Score', 'Runners-up', 'Country.1',
       'Venue', 'Attend­ance', 'Notes'],
      dtype='object')

In [166]:
df.index

RangeIndex(start=0, stop=70, step=1)

In [167]:
df.shape

(70, 9)

In [168]:
df['Notes'].unique()

array([nan, 'Match was won during extra time',
       'Match was won after a replay',
       'Match was won on a penalty shoot-out'], dtype=object)

In [169]:
df[df['Notes']=='Match was won after a replay']

Season       Country        Winners Score       Runners-up Country.1  \
18  1973–74  West Germany  Bayern Munich   1–1  Atlético Madrid     Spain   
19  1973–74  West Germany  Bayern Munich   4–0  Atlético Madrid     Spain   

                                Venue Attend­ance  \
18  Heysel Stadium, Brussels, Belgium      48,722   
19  Heysel Stadium, Brussels, Belgium      23,325   

                           Notes  
18  Match was won after a replay  
19  Match was won after a replay

In [170]:
df.drop([18],axis=0,inplace=True)

In [171]:
df[df['Notes']=='Match was won after a replay']

Season       Country        Winners Score       Runners-up Country.1  \
19  1973–74  West Germany  Bayern Munich   4–0  Atlético Madrid     Spain   

                                Venue Attend­ance  \
19  Heysel Stadium, Brussels, Belgium      23,325   

                           Notes  
19  Match was won after a replay

In [172]:
df['Season'].unique()

array(['1955–56', '1956–57', '1957–58', '1958–59', '1959–60', '1960–61',
       '1961–62', '1962–63', '1963–64', '1964–65', '1965–66', '1966–67',
       '1967–68', '1968–69', '1969–70', '1970–71', '1971–72', '1972–73',
       '1973–74', '1974–75', '1975–76', '1976–77', '1977–78', '1978–79',
       '1979–80', '1980–81', '1981–82', '1982–83', '1983–84', '1984–85',
       '1985–86', '1986–87', '1987–88', '1988–89', '1989–90', '1990–91',
       '1991–92', '1992–93', '1993–94', '1994–95', '1995–96', '1996–97',
       '1997–98', '1998–99', '1999–2000', '2000–01', '2001–02', '2002–03',
       '2003–04', '2004–05', '2005–06', '2006–07', '2007–08', '2008–09',
       '2009–10', '2010–11', '2011–12', '2012–13', '2013–14', '2014–15',
       '2015–16', '2016–17', '2017–18', '2018–19', '2019–20', '2020–21',
       '2021–22', '2022–23', '2023-24'], dtype=object)

In [173]:
df['Season'].nunique()

69

### The UCL Started 69 years ago 

In [174]:
df['Attend­ance'] = df['Attend­ance'].str.replace(',','')

In [175]:
df

Season  Country          Winners Score           Runners-up  \
0   1955–56    Spain      Real Madrid   4–3                Reims   
1   1956–57    Spain      Real Madrid   2–0           Fiorentina   
2   1957–58    Spain      Real Madrid   3–2                Milan   
3   1958–59    Spain      Real Madrid   2–0                Reims   
4   1959–60    Spain      Real Madrid   7–3  Eintracht Frankfurt   
..      ...      ...              ...   ...                  ...   
65  2019–20  Germany    Bayern Munich   1–0  Paris Saint-Germain   
66  2020–21  England          Chelsea   1–0      Manchester City   
67  2021–22    Spain      Real Madrid   1–0            Liverpool   
68  2022–23  England  Manchester City   1–0          Inter Milan   
69  2023-24    Spain      Real Madrid   2–0    Borussia Dortmund   

       Country.1                                      Venue Attend­ance  \
0         France            Parc des Princes, Paris, France       38239   
1          Italy           Santiago Bernabéu, Madrid, Spain      124000   
2          Italy          Heysel Stadium, Brussels, Belgium       67000   
3         France     Neckarstadion, Stuttgart, West Germany       72000   
4   West Germany            Hampden Park, Glasgow, Scotland      127621   
..           ...                                        ...         ...   
65        France           Estádio da Luz, Lisbon, Portugal           0   
66       England         Estádio do Dragão, Porto, Portugal       14110   
67       England       Stade de France, Saint-Denis, France       75000   
68         Italy  Atatürk Olympic Stadium, Istanbul, Turkey       71412   
69       Germany           Wembley Stadium, London, England       86212   

                              Notes  
0                               NaN  
1                               NaN  
2   Match was won during extra time  
3                               NaN  
4                               NaN  
..                              ...  
65                              NaN  
66                              NaN  
67                              NaN  
68                              NaN  
69                              NaN  

[69 rows x 9 columns]

In [176]:
df['Attend­ance'] = df['Attend­ance'].astype(int)

In [177]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 69 entries, 0 to 69
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Season       69 non-null     object
 1   Country      69 non-null     object
 2   Winners      69 non-null     object
 3   Score        69 non-null     object
 4   Runners-up   69 non-null     object
 5   Country.1    69 non-null     object
 6   Venue        69 non-null     object
 7   Attend­ance  69 non-null     int32 
 8   Notes        17 non-null     object
dtypes: int32(1), object(8)
memory usage: 5.1+ KB


In [178]:
df['Winners'].value_counts()

Winners
Real Madrid          15
Milan                 7
Bayern Munich         6
Liverpool             6
Barcelona             5
Ajax                  4
Inter Milan           3
Manchester United     3
Chelsea               2
Nottingham Forest     2
Benfica               2
Juventus              2
Porto                 2
Red Star Belgrade     1
Borussia Dortmund     1
Marseille             1
Aston Villa           1
PSV Eindhoven         1
Steaua București      1
Hamburger SV          1
Feyenoord             1
Celtic                1
Manchester City       1
Name: count, dtype: int64

In [179]:
df[df['Winners']=='Real Madrid']

Season Country      Winners Score           Runners-up     Country.1  \
0     1955–56   Spain  Real Madrid   4–3                Reims        France   
1     1956–57   Spain  Real Madrid   2–0           Fiorentina         Italy   
2     1957–58   Spain  Real Madrid   3–2                Milan         Italy   
3     1958–59   Spain  Real Madrid   2–0                Reims        France   
4     1959–60   Spain  Real Madrid   7–3  Eintracht Frankfurt  West Germany   
10    1965–66   Spain  Real Madrid   2–1             Partizan    Yugoslavia   
43    1997–98   Spain  Real Madrid   1–0             Juventus         Italy   
45  1999–2000   Spain  Real Madrid   3–0             Valencia         Spain   
47    2001–02   Spain  Real Madrid   2–1     Bayer Leverkusen       Germany   
59    2013–14   Spain  Real Madrid   4–1      Atlético Madrid         Spain   
61    2015–16   Spain  Real Madrid   1–1      Atlético Madrid         Spain   
62    2016–17   Spain  Real Madrid   4–1             Juventus         Italy   
63    2017–18   Spain  Real Madrid   3–1            Liverpool       England   
67    2021–22   Spain  Real Madrid   1–0            Liverpool       England   
69    2023-24   Spain  Real Madrid   2–0    Borussia Dortmund       Germany   

                                      Venue  Attend­ance  \
0           Parc des Princes, Paris, France        38239   
1          Santiago Bernabéu, Madrid, Spain       124000   
2         Heysel Stadium, Brussels, Belgium        67000   
3    Neckarstadion, Stuttgart, West Germany        72000   
4           Hampden Park, Glasgow, Scotland       127621   
10        Heysel Stadium, Brussels, Belgium        46745   
43  Amsterdam Arena, Amsterdam, Netherlands        48500   
45     Stade de France, Saint-Denis, France        80000   
47          Hampden Park, Glasgow, Scotland        50499   
59         Estádio da Luz, Lisbon, Portugal        60976   
61                   San Siro, Milan, Italy        71942   
62       Millennium Stadium, Cardiff, Wales        65842   
63   NSC Olimpiyskiy Stadium, Kyiv, Ukraine        61561   
67     Stade de France, Saint-Denis, France        75000   
69         Wembley Stadium, London, England        86212   

                                   Notes  
0                                    NaN  
1                                    NaN  
2        Match was won during extra time  
3                                    NaN  
4                                    NaN  
10                                   NaN  
43                                   NaN  
45                                   NaN  
47                                   NaN  
59       Match was won during extra time  
61  Match was won on a penalty shoot-out  
62                                   NaN  
63                                   NaN  
67                                   NaN  
69                                   NaN

## The most team won the final is Real Madrid with 15 titles

In [180]:
df['Runners-up'].value_counts()

Runners-up
Juventus                    7
Bayern Munich               5
Benfica                     5
Liverpool                   4
Milan                       4
Barcelona                   3
Real Madrid                 3
Inter Milan                 3
Atlético Madrid             3
Valencia                    2
Reims                       2
Borussia Dortmund           2
Ajax                        2
Manchester United           2
Steaua București            1
Chelsea                     1
Arsenal                     1
Monaco                      1
Bayer Leverkusen            1
Tottenham Hotspur           1
Sampdoria                   1
Paris Saint-Germain         1
Marseille                   1
Malmö FF                    1
Roma                        1
Hamburger SV                1
Fiorentina                  1
Club Brugge                 1
Borussia Mönchengladbach    1
Saint-Étienne               1
Leeds United                1
Panathinaikos               1
Celtic                      1

In [181]:
df[df['Runners-up']=='Juventus']

Season       Country            Winners Score Runners-up Country.1  \
17  1972–73   Netherlands               Ajax   1–0   Juventus     Italy   
28  1982–83  West Germany       Hamburger SV   1–0   Juventus     Italy   
42  1996–97       Germany  Borussia Dortmund   3–1   Juventus     Italy   
43  1997–98         Spain        Real Madrid   1–0   Juventus     Italy   
48  2002–03         Italy              Milan   0–0   Juventus     Italy   
60  2014–15         Spain          Barcelona   3–1   Juventus     Italy   
62  2016–17         Spain        Real Madrid   4–1   Juventus     Italy   

                                         Venue  Attend­ance  \
17  Red Star Stadium, Belgrade, SFR Yugoslavia        89484   
28             Olympic Stadium, Athens, Greece        73500   
42             Olympiastadion, Munich, Germany        59000   
43     Amsterdam Arena, Amsterdam, Netherlands        48500   
48           Old Trafford, Manchester, England        62315   
60             Olympiastadion, Berlin, Germany        70442   
62          Millennium Stadium, Cardiff, Wales        65842   

                                   Notes  
17                                   NaN  
28                                   NaN  
42                                   NaN  
43                                   NaN  
48  Match was won on a penalty shoot-out  
60                                   NaN  
62                                   NaN

## The most team lost the final is Juventus with 7 times

In [182]:
df['Country'].value_counts()

Country
Spain           20
England         15
Italy           12
Netherlands      6
Portugal         4
West Germany     4
Germany          4
Scotland         1
Romania          1
Yugoslavia       1
France           1
Name: count, dtype: int64

## The most country won the final is sapin with 20 titles

In [183]:
df['Country.1'].value_counts()

Country.1
Italy           17
Spain           11
England         11
France           6
Germany          6
West Germany     5
Portugal         5
Netherlands      2
Yugoslavia       1
Scotland         1
Greece           1
Belgium          1
Sweden           1
Romania          1
Name: count, dtype: int64

## The most country lost the final is italy with 17 times

## The most country reach the final is Sapin with 31 times

In [184]:
df['Score'].value_counts()

Score
1–0    19
2–1    10
2–0     9
3–1     6
1–1     6
4–1     4
0–0     4
4–0     3
3–2     2
3–0     2
4–3     1
7–3     1
5–3     1
3–3     1
Name: count, dtype: int64

## The most score is repeated is 1-0 for 19 times

## Draw  is repeated  11 times

## The negative draw  is repeated  4 times 

## The positve  draw  is repeated  7 times 

In [185]:
df[df['Score']==df['Score'].max()]

Season Country      Winners Score           Runners-up     Country.1  \
4  1959–60   Spain  Real Madrid   7–3  Eintracht Frankfurt  West Germany   

                             Venue  Attend­ance Notes  
4  Hampden Park, Glasgow, Scotland       127621   NaN

#### The most goal-filled match was the  in sesaon 1959-60 between Real Madrid and Eintracht Frankfurt with 10 golas 7-3 for Real Madrid


In [186]:
df['Notes'].value_counts()

Notes
Match was won on a penalty shoot-out    11
Match was won during extra time          5
Match was won after a replay             1
Name: count, dtype: int64

## Match was won on a penalty shoot-out in 11 times

In [187]:
df[df['Notes']=='Match was won on a penalty shoot-out'].iloc[-1]['Season']

'2015–16'

In [188]:
df[df['Notes']=='Match was won on a penalty shoot-out'].tail(1)

Season Country      Winners Score       Runners-up Country.1  \
61  2015–16   Spain  Real Madrid   1–1  Atlético Madrid     Spain   

                     Venue  Attend­ance                                 Notes  
61  San Siro, Milan, Italy        71942  Match was won on a penalty shoot-out

In [189]:
df[df['Notes']=='Match was won on a penalty shoot-out'].tail(1)['Season']

61    2015–16
Name: Season, dtype: object

In [190]:
df[df['Notes']=='Match was won on a penalty shoot-out'].head(1)

Season  Country    Winners Score Runners-up Country.1  \
29  1983–84  England  Liverpool   1–1       Roma     Italy   

                           Venue  Attend­ance  \
29  Stadio Olimpico, Rome, Italy        69693   

                                   Notes  
29  Match was won on a penalty shoot-out

## The first time the penalties finish the final  in 1984 with Liverpool 

## The last time the penalties finish the final  in 2016 with Real Madrid 

In [191]:
df['Notes'].value_counts()

Notes
Match was won on a penalty shoot-out    11
Match was won during extra time          5
Match was won after a replay             1
Name: count, dtype: int64

In [192]:
df[df['Notes']=='Match was won during extra time']

Season      Country            Winners Score       Runners-up Country.1  \
2   1957–58        Spain        Real Madrid   3–2            Milan     Italy   
12  1967–68      England  Manchester United   4–1          Benfica  Portugal   
14  1969–70  Netherlands          Feyenoord   2–1           Celtic  Scotland   
37  1991–92        Spain          Barcelona   1–0        Sampdoria     Italy   
59  2013–14        Spain        Real Madrid   4–1  Atlético Madrid     Spain   

                                Venue  Attend­ance  \
2   Heysel Stadium, Brussels, Belgium        67000   
12   Wembley Stadium, London, England        92225   
14             San Siro, Milan, Italy        53187   
37   Wembley Stadium, London, England        70827   
59   Estádio da Luz, Lisbon, Portugal        60976   

                              Notes  
2   Match was won during extra time  
12  Match was won during extra time  
14  Match was won during extra time  
37  Match was won during extra time  
59  Match was won during extra time

In [193]:
len(df[df['Notes']=='Match was won during extra time'])

5

## Match was won during extra time  in 5 times

In [194]:
df['Venue'].value_counts()

Venue
Wembley Stadium, London, England              8
Stadio Olimpico, Rome, Italy                  4
Heysel Stadium, Brussels, Belgium             4
San Siro, Milan, Italy                        4
Santiago Bernabéu, Madrid, Spain              4
Parc des Princes, Paris, France               3
Hampden Park, Glasgow, Scotland               3
Stade de France, Saint-Denis, France          3
Praterstadion, Vienna, Austria                3
Olympic Stadium, Athens, Greece               3
Estádio da Luz, Lisbon, Portugal              2
Atatürk Olympic Stadium, Istanbul, Turkey     2
Olympiastadion, Munich, Germany               2
Camp Nou, Barcelona, Spain                    2
De Kuip, Rotterdam, Netherlands               2
Neckarstadion, Stuttgart, West Germany        2
Olympiastadion, Munich, West Germany          1
Wankdorf Stadium, Bern, Switzerland           1
Metropolitano Stadium, Madrid, Spain          1
NSC Olimpiyskiy Stadium, Kyiv, Ukraine        1
Millennium Stadium, Cardiff, Wales

In [195]:
df[df['Venue']=='Wembley Stadium, London, England']

Season      Country            Winners Score         Runners-up  \
7   1962–63        Italy              Milan   2–1            Benfica   
12  1967–68      England  Manchester United   4–1            Benfica   
15  1970–71  Netherlands               Ajax   2–0      Panathinaikos   
23  1977–78      England          Liverpool   1–0        Club Brugge   
37  1991–92        Spain          Barcelona   1–0          Sampdoria   
56  2010–11        Spain          Barcelona   3–1  Manchester United   
58  2012–13      Germany      Bayern Munich   2–1  Borussia Dortmund   
69  2023-24        Spain        Real Madrid   2–0  Borussia Dortmund   

   Country.1                             Venue  Attend­ance  \
7   Portugal  Wembley Stadium, London, England        45715   
12  Portugal  Wembley Stadium, London, England        92225   
15    Greece  Wembley Stadium, London, England        83179   
23   Belgium  Wembley Stadium, London, England        92500   
37     Italy  Wembley Stadium, London, England        70827   
56   England  Wembley Stadium, London, England        87695   
58   Germany  Wembley Stadium, London, England        86298   
69   Germany  Wembley Stadium, London, England        86212   

                              Notes  
7                               NaN  
12  Match was won during extra time  
15                              NaN  
23                              NaN  
37  Match was won during extra time  
56                              NaN  
58                              NaN  
69                              NaN

## The most Stadium have the final is Wembley in London

In [196]:
df['Attend­ance'].max()

127621

In [197]:
df[df['Attend­ance'] == df['Attend­ance'].max()]

Season Country      Winners Score           Runners-up     Country.1  \
4  1959–60   Spain  Real Madrid   7–3  Eintracht Frankfurt  West Germany   

                             Venue  Attend­ance Notes  
4  Hampden Park, Glasgow, Scotland       127621   NaN

### The final has a most attend­ance in season 1959-60 in West-Germany it was 127,621

In [198]:
df[df['Attend­ance'] == df['Attend­ance'].min()]

Season  Country        Winners Score           Runners-up Country.1  \
65  2019–20  Germany  Bayern Munich   1–0  Paris Saint-Germain    France   

                               Venue  Attend­ance Notes  
65  Estádio da Luz, Lisbon, Portugal            0   NaN

### The final has  least attend­ance in season 2019-20 in Portugal  it was 0 because covid-19

In [199]:
def count_golas(s): 
    s = s.split('–')
    return int(s[0]) + int(s[1])


In [200]:
df['Counter_goals'] = df['Score'].apply(count_golas)

In [201]:
df

Season  Country          Winners Score           Runners-up  \
0   1955–56    Spain      Real Madrid   4–3                Reims   
1   1956–57    Spain      Real Madrid   2–0           Fiorentina   
2   1957–58    Spain      Real Madrid   3–2                Milan   
3   1958–59    Spain      Real Madrid   2–0                Reims   
4   1959–60    Spain      Real Madrid   7–3  Eintracht Frankfurt   
..      ...      ...              ...   ...                  ...   
65  2019–20  Germany    Bayern Munich   1–0  Paris Saint-Germain   
66  2020–21  England          Chelsea   1–0      Manchester City   
67  2021–22    Spain      Real Madrid   1–0            Liverpool   
68  2022–23  England  Manchester City   1–0          Inter Milan   
69  2023-24    Spain      Real Madrid   2–0    Borussia Dortmund   

       Country.1                                      Venue  Attend­ance  \
0         France            Parc des Princes, Paris, France        38239   
1          Italy           Santiago Bernabéu, Madrid, Spain       124000   
2          Italy          Heysel Stadium, Brussels, Belgium        67000   
3         France     Neckarstadion, Stuttgart, West Germany        72000   
4   West Germany            Hampden Park, Glasgow, Scotland       127621   
..           ...                                        ...          ...   
65        France           Estádio da Luz, Lisbon, Portugal            0   
66       England         Estádio do Dragão, Porto, Portugal        14110   
67       England       Stade de France, Saint-Denis, France        75000   
68         Italy  Atatürk Olympic Stadium, Istanbul, Turkey        71412   
69       Germany           Wembley Stadium, London, England        86212   

                              Notes  Counter_goals  
0                               NaN              7  
1                               NaN              2  
2   Match was won during extra time              5  
3                               NaN              2  
4                               NaN             10  
..                              ...            ...  
65                              NaN              1  
66                              NaN              1  
67                              NaN              1  
68                              NaN              1  
69                              NaN              2  

[69 rows x 10 columns]

In [202]:
def winners_goals(s): 
    s = s.split('–')
    return int(s[0])

In [203]:
df["Winners_goals"] = df['Score'].apply(winners_goals)

In [204]:
def losers_goals(s): 
    s = s.split('–')
    return int(s[1])

In [205]:
df["losers_goals"] = df['Score'].apply(losers_goals)

In [206]:
df

Season  Country          Winners Score           Runners-up  \
0   1955–56    Spain      Real Madrid   4–3                Reims   
1   1956–57    Spain      Real Madrid   2–0           Fiorentina   
2   1957–58    Spain      Real Madrid   3–2                Milan   
3   1958–59    Spain      Real Madrid   2–0                Reims   
4   1959–60    Spain      Real Madrid   7–3  Eintracht Frankfurt   
..      ...      ...              ...   ...                  ...   
65  2019–20  Germany    Bayern Munich   1–0  Paris Saint-Germain   
66  2020–21  England          Chelsea   1–0      Manchester City   
67  2021–22    Spain      Real Madrid   1–0            Liverpool   
68  2022–23  England  Manchester City   1–0          Inter Milan   
69  2023-24    Spain      Real Madrid   2–0    Borussia Dortmund   

       Country.1                                      Venue  Attend­ance  \
0         France            Parc des Princes, Paris, France        38239   
1          Italy           Santiago Bernabéu, Madrid, Spain       124000   
2          Italy          Heysel Stadium, Brussels, Belgium        67000   
3         France     Neckarstadion, Stuttgart, West Germany        72000   
4   West Germany            Hampden Park, Glasgow, Scotland       127621   
..           ...                                        ...          ...   
65        France           Estádio da Luz, Lisbon, Portugal            0   
66       England         Estádio do Dragão, Porto, Portugal        14110   
67       England       Stade de France, Saint-Denis, France        75000   
68         Italy  Atatürk Olympic Stadium, Istanbul, Turkey        71412   
69       Germany           Wembley Stadium, London, England        86212   

                              Notes  Counter_goals  Winners_goals  \
0                               NaN              7              4   
1                               NaN              2              2   
2   Match was won during extra time              5              3   
3                               NaN              2              2   
4                               NaN             10              7   
..                              ...            ...            ...   
65                              NaN              1              1   
66                              NaN              1              1   
67                              NaN              1              1   
68                              NaN              1              1   
69                              NaN              2              2   

    losers_goals  
0              3  
1              0  
2              2  
3              0  
4              3  
..           ...  
65             0  
66             0  
67             0  
68             0  
69             0  

[69 rows x 12 columns]

In [207]:
all_winners = df['Winners'].unique()

In [208]:
all_runners = df['Runners-up'].unique()

In [209]:
all_teams = np.append(all_winners,all_runners)

In [210]:
teams = set(all_teams)

In [211]:
teams = np.array(list(teams))
teams

array(['Eintracht Frankfurt', 'Atlético Madrid', 'Bayer Leverkusen',
       'Nottingham Forest', 'Club Brugge', 'Monaco',
       'Paris Saint-Germain', 'Inter Milan', 'Roma', 'Aston Villa',
       'Red Star Belgrade', 'Saint-Étienne', 'Celtic',
       'Manchester United', 'Chelsea', 'Real Madrid', 'Reims', 'Malmö FF',
       'Partizan', 'Milan', 'Benfica', 'Borussia Dortmund', 'Juventus',
       'Tottenham Hotspur', 'Arsenal', 'Liverpool', 'Marseille', 'Porto',
       'Barcelona', 'Leeds United', 'Hamburger SV', 'Steaua București',
       'Fiorentina', 'Panathinaikos', 'Ajax', 'Valencia', 'Feyenoord',
       'Sampdoria', 'Bayern Munich', 'PSV Eindhoven',
       'Borussia Mönchengladbach', 'Manchester City'], dtype='<U24')

In [212]:
len(teams)

42

In [213]:
teams_golas={}
def goals_scored():
    for i in teams :
        h = df[df['Winners']==i]['Winners_goals'].sum()
        a = df[df['Runners-up']==i]['losers_goals'].sum()
        teams_golas.update({i:h+a})
goals_scored()

In [214]:
teams_golas

{'Eintracht Frankfurt': 3,
 'Atlético Madrid': 2,
 'Bayer Leverkusen': 1,
 'Nottingham Forest': 2,
 'Club Brugge': 0,
 'Monaco': 0,
 'Paris Saint-Germain': 0,
 'Inter Milan': 7,
 'Roma': 1,
 'Aston Villa': 1,
 'Red Star Belgrade': 0,
 'Saint-Étienne': 0,
 'Celtic': 3,
 'Manchester United': 8,
 'Chelsea': 3,
 'Real Madrid': 45,
 'Reims': 3,
 'Malmö FF': 0,
 'Partizan': 1,
 'Milan': 22,
 'Benfica': 10,
 'Borussia Dortmund': 4,
 'Juventus': 5,
 'Tottenham Hotspur': 0,
 'Arsenal': 1,
 'Liverpool': 13,
 'Marseille': 1,
 'Porto': 5,
 'Barcelona': 13,
 'Leeds United': 0,
 'Hamburger SV': 1,
 'Steaua București': 0,
 'Fiorentina': 0,
 'Panathinaikos': 0,
 'Ajax': 8,
 'Valencia': 1,
 'Feyenoord': 2,
 'Sampdoria': 0,
 'Bayern Munich': 14,
 'PSV Eindhoven': 0,
 'Borussia Mönchengladbach': 1,
 'Manchester City': 1}

In [215]:
a = sorted(teams_golas.items(), key=lambda x: x[1],reverse=True)
a

[('Real Madrid', 45),
 ('Milan', 22),
 ('Bayern Munich', 14),
 ('Liverpool', 13),
 ('Barcelona', 13),
 ('Benfica', 10),
 ('Manchester United', 8),
 ('Ajax', 8),
 ('Inter Milan', 7),
 ('Juventus', 5),
 ('Porto', 5),
 ('Borussia Dortmund', 4),
 ('Eintracht Frankfurt', 3),
 ('Celtic', 3),
 ('Chelsea', 3),
 ('Reims', 3),
 ('Atlético Madrid', 2),
 ('Nottingham Forest', 2),
 ('Feyenoord', 2),
 ('Bayer Leverkusen', 1),
 ('Roma', 1),
 ('Aston Villa', 1),
 ('Partizan', 1),
 ('Arsenal', 1),
 ('Marseille', 1),
 ('Hamburger SV', 1),
 ('Valencia', 1),
 ('Borussia Mönchengladbach', 1),
 ('Manchester City', 1),
 ('Club Brugge', 0),
 ('Monaco', 0),
 ('Paris Saint-Germain', 0),
 ('Red Star Belgrade', 0),
 ('Saint-Étienne', 0),
 ('Malmö FF', 0),
 ('Tottenham Hotspur', 0),
 ('Leeds United', 0),
 ('Steaua București', 0),
 ('Fiorentina', 0),
 ('Panathinaikos', 0),
 ('Sampdoria', 0),
 ('PSV Eindhoven', 0)]

### The most team scored goals is Real Madrid with 45 goals

In [216]:
get_goals_teams = {}
def get_goals(): 
    for i in teams: 
        h = df[df['Winners']==i]['losers_goals'].sum()
        a = df[df['Runners-up']==i]['Winners_goals'].sum()
        get_goals_teams.update({i:h+a})
get_goals()

In [217]:
get_goals_teams

{'Eintracht Frankfurt': 7,
 'Atlético Madrid': 9,
 'Bayer Leverkusen': 2,
 'Nottingham Forest': 0,
 'Club Brugge': 1,
 'Monaco': 3,
 'Paris Saint-Germain': 1,
 'Inter Milan': 6,
 'Roma': 1,
 'Aston Villa': 0,
 'Red Star Belgrade': 0,
 'Saint-Étienne': 1,
 'Celtic': 3,
 'Manchester United': 8,
 'Chelsea': 2,
 'Real Madrid': 23,
 'Reims': 6,
 'Malmö FF': 1,
 'Partizan': 2,
 'Milan': 11,
 'Benfica': 13,
 'Borussia Dortmund': 5,
 'Juventus': 14,
 'Tottenham Hotspur': 2,
 'Arsenal': 2,
 'Liverpool': 12,
 'Marseille': 0,
 'Porto': 1,
 'Barcelona': 10,
 'Leeds United': 2,
 'Hamburger SV': 1,
 'Steaua București': 4,
 'Fiorentina': 2,
 'Panathinaikos': 2,
 'Ajax': 5,
 'Valencia': 4,
 'Feyenoord': 1,
 'Sampdoria': 1,
 'Bayern Munich': 10,
 'PSV Eindhoven': 0,
 'Borussia Mönchengladbach': 3,
 'Manchester City': 1}

In [218]:
b = sorted(get_goals_teams.items(), key=lambda x: x[1],reverse=True)
b

[('Real Madrid', 23),
 ('Juventus', 14),
 ('Benfica', 13),
 ('Liverpool', 12),
 ('Milan', 11),
 ('Barcelona', 10),
 ('Bayern Munich', 10),
 ('Atlético Madrid', 9),
 ('Manchester United', 8),
 ('Eintracht Frankfurt', 7),
 ('Inter Milan', 6),
 ('Reims', 6),
 ('Borussia Dortmund', 5),
 ('Ajax', 5),
 ('Steaua București', 4),
 ('Valencia', 4),
 ('Monaco', 3),
 ('Celtic', 3),
 ('Borussia Mönchengladbach', 3),
 ('Bayer Leverkusen', 2),
 ('Chelsea', 2),
 ('Partizan', 2),
 ('Tottenham Hotspur', 2),
 ('Arsenal', 2),
 ('Leeds United', 2),
 ('Fiorentina', 2),
 ('Panathinaikos', 2),
 ('Club Brugge', 1),
 ('Paris Saint-Germain', 1),
 ('Roma', 1),
 ('Saint-Étienne', 1),
 ('Malmö FF', 1),
 ('Porto', 1),
 ('Hamburger SV', 1),
 ('Feyenoord', 1),
 ('Sampdoria', 1),
 ('Manchester City', 1),
 ('Nottingham Forest', 0),
 ('Aston Villa', 0),
 ('Red Star Belgrade', 0),
 ('Marseille', 0),
 ('PSV Eindhoven', 0)]

### The most team get goals is Real Madrid with 23 goals

In [219]:
def win_percentage():
    for i in teams:
        w = df[df['Winners']==i].shape[0]
        l = df[df['Runners-up']==i].shape[0]
        print(f"{i} reach the final: {w+l},  won in: {w}  and lost in: {l}  percntage to win  the ucl is {w/(w+l)*100}")
        print()
win_percentage()

Eintracht Frankfurt reach the final: 1,  won in: 0  and lost in: 1  percntage to win  the ucl is 0.0

Atlético Madrid reach the final: 3,  won in: 0  and lost in: 3  percntage to win  the ucl is 0.0

Bayer Leverkusen reach the final: 1,  won in: 0  and lost in: 1  percntage to win  the ucl is 0.0

Nottingham Forest reach the final: 2,  won in: 2  and lost in: 0  percntage to win  the ucl is 100.0

Club Brugge reach the final: 1,  won in: 0  and lost in: 1  percntage to win  the ucl is 0.0

Monaco reach the final: 1,  won in: 0  and lost in: 1  percntage to win  the ucl is 0.0

Paris Saint-Germain reach the final: 1,  won in: 0  and lost in: 1  percntage to win  the ucl is 0.0

Inter Milan reach the final: 6,  won in: 3  and lost in: 3  percntage to win  the ucl is 50.0

Roma reach the final: 1,  won in: 0  and lost in: 1  percntage to win  the ucl is 0.0

Aston Villa reach the final: 1,  won in: 1  and lost in: 0  percntage to win  the ucl is 100.0

Red Star Belgrade reach the final: 1